Include/cpython/pyatomic.h

```diff
@@ -0,0 +1,397 @@
+#ifndef Py_ATOMIC_H
+#define Py_ATOMIC_H
+
+// This header provides cross-platform low-level atomic operations
+// similar to C11 atomics.
+//
+// Operations are sequentially consistent unless they have a suffix indicating
+// otherwise. If in doubt, prefer the sequentially consistent operations.
+//
+// The "_relaxed" suffix for load and store operations indicates the "relaxed"
+// memory order. They don't provide synchronization, but (roughly speaking)
+// guarantee somewhat sane behavior for races instead of undefined behavior.
+// In practice, they correspond to "normal" hardware load and store instructions,
+// so they are almost as inexpensive as plain loads and stores in C.
+//
+// Note that atomic read-modify-write operations like _Py_atomic_add_* return
+// the previous value of the atomic variable, not the new value.
+//
+// See https://en.cppreference.com/w/c/atomic for more information on C11 atomics.
+// See https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2055r0.pdf
+// "A Relaxed Guide to memory_order_relaxed" for discussion of and common usage
+// or relaxed atomics.
+
+// Atomically adds `value` to `address` and returns the previous value
+static inline int
+_Py_atomic_add_int(int *address, int value);
+
+static inline int8_t
+_Py_atomic_add_int8(int8_t *address, int8_t value);
+
+static inline int16_t
+_Py_atomic_add_int16(int16_t *address, int16_t value);
+
+static inline int32_t
+_Py_atomic_add_int32(int32_t *address, int32_t value);
+
+static inline int64_t
+_Py_atomic_add_int64(int64_t *address, int64_t value);
+
+static inline intptr_t
+_Py_atomic_add_intptr(intptr_t *address, intptr_t value);
+
+static inline unsigned int
+_Py_atomic_add_uint(unsigned int *address, unsigned int value);
+
+static inline uint8_t
+_Py_atomic_add_uint8(uint8_t *address, uint8_t value);
+
+static inline uint16_t
+_Py_atomic_add_uint16(uint16_t *address, uint16_t value);
+
+static inline uint32_t
+_Py_atomic_add_uint32(uint32_t *address, uint32_t value);
+
+static inline uint64_t
+_Py_atomic_add_uint64(uint64_t *address, uint64_t value);
+
+static inline uintptr_t
+_Py_atomic_add_uintptr(uintptr_t *address, uintptr_t value);
+
+static inline Py_ssize_t
+_Py_atomic_add_ssize(Py_ssize_t *address, Py_ssize_t value);
```

Comment:
Not sure it's worth exposing atomic ops for all int sizes and signednesses?

Reply:
I expect to use at least one atomic operation on each of the data types here (but not every atomic op on every data type). I tried to be consistent on what's defined because it makes understanding what's available easier and testing easier.